### ISSUE TITLE
Detect rectangles in black and white image - custom dataset

### ISSUE THREAD

**Issue by elliestath** [matterport/Mask_RCNN#2937]
Hello, I want to detect some random rectangle shapes on thresholded (black&white) images. 
I have only one class ("rectangle").
My images look like this:

![example](https://user-images.githubusercontent.com/11046400/225887426-443449f9-c615-4ba6-96c4-e3d8402ac094.jpg)

I have 142 annotated images (VIA format) and I use 119 for training and 23 for validation.
Image size is 512x512.
Given the small size of my dataset, I have also created a synthetic dataset of the same size 512 x 512 (let's say of 1000 (train)/100 (val) images, but can easily be more) with rectangle shapes based on the scripts provided in shapes.py. Rectangles and respective masks look like this:

![Screenshot from 2023-03-17 12-08-00](https://user-images.githubusercontent.com/11046400/225888563-c652af38-fc72-411a-9f16-4a793bfa4fcb.png)

Following the examples, I trained my synthetic dataset using the following config:

```
class RectanglesConfig(Config):
    """Configuration for training on the toy shapes dataset.
    Derives from the base Config class and overrides values specific
    to the toy shapes dataset.
    """
    # Give the configuration a recognizable name
    NAME = "rectangles"

    # Train on 1 GPU and 8 images per GPU. We can put multiple images on each
    # GPU because the images are small. Batch size is 8 (GPUs * images/GPU).
    GPU_COUNT = 1
    IMAGES_PER_GPU = 8

    # Number of classes (including background)
    NUM_CLASSES = 1 + 1  # background + 3 shapes

    # Use small images for faster training. Set the limits of the small side
    # the large side, and that determines the image shape.
    IMAGE_MIN_DIM = 512
    IMAGE_MAX_DIM = 512

    # Use smaller anchors because our image and objects are small
    RPN_ANCHOR_SCALES = (8, 16, 32, 64, 128)  # anchor side in pixels

    # Reduce training ROIs per image because the images are small and have
    # few objects. Aim to allow ROI sampling to pick 33% positive ROIs.
    TRAIN_ROIS_PER_IMAGE = 32

    # Use a small epoch since the data is simple
    STEPS_PER_EPOCH = 100

    # use small validation steps since the epoch is small
    VALIDATION_STEPS = 5
    
config = RectanglesConfig()
config.display()
```
I used data augmentation like that:
```
augmentation = imgaug.augmenters.Sometimes(0.5, imgaug.augmenters.Sequential([imgaug.augmenters.Fliplr(0.5),
                                                                         imgaug.augmenters.Flipud(0.5),
                                                                         imgaug.augmenters.Affine(rotate=(0, 90)),
                                                                         imgaug.augmenters.Affine(scale=(0.8, 1.5)),
                                                                         imgaug.augmenters.AdditiveLaplaceNoise(scale=(0, 0.3*255))
                                                                        ]))

image, image_meta, class_ids, bbox, mask = modellib.load_image_gt(
    dataset_train, config, image_id, augmentation=augmentation, use_mini_mask=False)
log("mask", mask)
display_images([image]+[mask[:,:,i] for i in range(min(mask.shape[-1], 7))])
```

I had to use `resnet50` due to GPU limitations. 
I use multi-step training starting from COCO weights and a learning rate of 0.0001 like this:

```
model.train(dataset_train, dataset_val, 
            learning_rate=config.LEARNING_RATE, 
            epochs=30, 
            layers='heads',
            augmentation=augmentation)
```

and then:

```
model.train(dataset_train, dataset_val, 
            learning_rate=config.LEARNING_RATE / 10,
            epochs=50, 
            layers="all",
            augmentation=augmentation)
```

I got satisfying loss curves for train ( <0.50)  and val (<0.30) and good inference results on a synthetic test dataset (ignore "footprint", it means "rectangle"):
![Screenshot from 2023-03-17 14-34-16](https://user-images.githubusercontent.com/11046400/225934861-7ff242a7-ea4f-4763-a4f4-83f7b0f982b1.png)

Then, starting from those weights of the synthetic dataset, I try to fine-tune on my real dataset of 119 (train)/23(val) images.
Using the following config (smaller batch size = 4 since I have fewer data):

```
class RectanglesConfig(Config):
    """Configuration for training on the custom footprint dataset.
    Derives from the base Config class and overrides values specific
    to the rectangle dataset.
    """
    # Give the configuration a recognizable name
    NAME = "rectangle"

    # Train on 1 GPU and 8 images per GPU. We can put multiple images on each
    # GPU because the images are small. Batch size is 8 (GPUs * images/GPU).
    GPU_COUNT = 1
    IMAGES_PER_GPU = 4

    # Number of classes (including background)
    NUM_CLASSES = 1 + 1  # background + 1 shape

    # Use small images for faster training. Set the limits of the small side
    # the large side, and that determines the image shape.
    IMAGE_MIN_DIM = 512
    IMAGE_MAX_DIM = 512

    # Use smaller anchors because our image and objects are small
    RPN_ANCHOR_SCALES = (64, 128, 256)  # anchor side in pixels

    # Reduce training ROIs per image because the images are small and have
    # few objects. Aim to allow ROI sampling to pick 33% positive ROIs.
    TRAIN_ROIS_PER_IMAGE = 32

    # Use a small epoch since the data is simple
    STEPS_PER_EPOCH = 100

    # use small validation steps since the epoch is small
    VALIDATION_STEPS = 5
    
    LEARNING_RATE = 0.0001
    
    MASK_SHAPE = [512, 512]
    
config = RectanglesConfig()
config.display()


```
Using again the same augmentation and learning rate 0.0001 to train only the heads:

```
model.train(dataset_train, dataset_val, 
            learning_rate=config.LEARNING_RATE, 
            epochs=30, 
            layers='heads',
            augmentation=augmentation)
```
With losses:

```
Epoch 30/30
100/100 [==============================] - 63s 633ms/step - loss: 1.3106 - rpn_class_loss: 0.0607 - rpn_bbox_loss: 0.5746 - mrcnn_class_loss: 0.1462 - mrcnn_bbox_loss: 0.2148 - mrcnn_mask_loss: 0.3143 - val_loss: 1.3441 - val_rpn_class_loss: 0.0902 - val_rpn_bbox_loss: 0.4253 - val_mrcnn_class_loss: 0.2028 - val_mrcnn_bbox_loss: 0.2760 - val_mrcnn_mask_loss: 0.3498
```

![loss](https://user-images.githubusercontent.com/11046400/225932886-447902ac-26af-4918-92a7-7975fb025557.png)

Then I fine-tune 

```
print("Fine tune Resnet stage 4 and up")
model.train(dataset_train, dataset_val,
            learning_rate=0.0001,
            epochs=50,
            layers='4+',
            augmentation=augmentation)
```

With losses:
```
Epoch 50/50
100/100 [==============================] - 68s 679ms/step - loss: 1.1970 - rpn_class_loss: 0.0493 - rpn_bbox_loss: 0.5257 - mrcnn_class_loss: 0.1176 - mrcnn_bbox_loss: 0.2032 - mrcnn_mask_loss: 0.3012 - val_loss: 1.2893 - val_rpn_class_loss: 0.0689 - val_rpn_bbox_loss: 0.4054 - val_mrcnn_class_loss: 0.1709 - val_mrcnn_bbox_loss: 0.2711 - val_mrcnn_mask_loss: 0.3730
```

And then I fine-tune all the layers:

```
model.train(dataset_train, dataset_val, 
            learning_rate=0.0001,
            epochs=80, 
            layers="all",
            augmentation=augmentation)
```

With losses:

```
Epoch 80/80
100/100 [==============================] - 83s 827ms/step - loss: 1.0324 - rpn_class_loss: 0.0394 - rpn_bbox_loss: 0.4612 - mrcnn_class_loss: 0.0815 - mrcnn_bbox_loss: 0.1726 - mrcnn_mask_loss: 0.2777 - val_loss: 1.3186 - val_rpn_class_loss: 0.0876 - val_rpn_bbox_loss: 0.3843 - val_mrcnn_class_loss: 0.1762 - val_mrcnn_bbox_loss: 0.2795 - val_mrcnn_mask_loss: 0.3911
```

During this last training stage, losses are not decreasing enough (train_loss fluctuates between 0.8-1.1, and val_loss is always around 1.20-1.30) and in consequence, my inference results are bad; few true positives, some false positives, and many many rectangles remain undetected. For example:

![fine-tuned](https://user-images.githubusercontent.com/11046400/225948603-86640e83-6e6c-46ea-8772-13c7a742527d.png)

What do you think is the problem? Too few real data? Wrong training strategy? Wrong hyperparameters? Inconsistency between synthetic and real data? Any hint would be appreciated! Thanks!
